qutip/control/pulsegen.py

```diff
@@ -234,50 +236,120 @@ def init_pulse(self):
 
         self._pulse_initialised = True
 
-        if not self.lbound is None:
-            if np.isinf(self.lbound):
-                self.lbound = None
-        if not self.ubound is None:
-            if np.isinf(self.ubound):
-                self.ubound = None
-
-        if not self.ubound is None and not self.lbound is None:
-            if self.ubound < self.lbound:
+        if (
+            isinstance(self.lbound, np.ndarray) or
+            isinstance(self.ubound, np.ndarray)
+        ):
+            if not (
+                isinstance(self.lbound, np.ndarray) and
+                isinstance(self.ubound, np.ndarray)
+            ):
+                raise ValueError(
+                    "If either lbound or ubound is an array, both must be."
+                )
+            if self.lbound.shape != self.ubound.shape:
+                raise ValueError(
+                    "If lbound and ubound are arrays, they should have the"
+                    " same shape"
+                )
+            if None in self.lbound or None in self.ubound:
+                raise ValueError(
+                    "If lbound and ubound are arrays, use np.inf or -np.inf,"
+                    " not None, to specify the lack of a bound."
+                )
```

Comment:
I think the bounds for the pulse generator and optimizer are handled completely separately and I can't see any code paths where `_build_bounds_list` would have been called on `lbound` or `ubound` here.

I think it is consistent, just odd. In both interfaces, arrays of bounds must contain `np.inf` and not `None` (which makes sense otherwise the dtype of the ndarray will have to be object, which will make the array bigger and slower).

Then for the lower level optimizer, the `np.inf` is converted to `None` (presumably for the underlying optimizer) but this predates this PR (I have no idea whether this step is actually necessary or beneficial).